save-orchestrator/src/main/kotlin/org/cqfn/save/orchestrator/controller/HeartbeatController.kt

```diff
@@ -82,18 +101,77 @@ class HeartbeatController(private val agentService: AgentService,
     private fun handleVacantAgent(agentId: String): Mono<HeartbeatResponse> = agentService.getNewTestsIds(agentId)
         .doOnSuccess {
             if (it is WaitResponse) {
-                initiateShutdownSequence(agentId)
+                initiateShutdownSequence(agentId, areAllAgentsCrashed = false)
             }
         }
 
+    /**
+     * Collect information about latest heartbeats from agents, in aim to determine crashed one later
+     *
+     * @param agentId
+     * @param state
+     */
+    fun updateAgentHeartbeatTimeStamps(agentId: String, state: AgentState) {
+        agentsLatestHeartBeatsMap[agentId] = state.name to LocalDateTime.now()
```

Comment:
I thought, we store timestamp on heartbeat sending in the agent... We should consider adding it in the future

Reply:
Open https://github.com/analysis-dev/save-cloud/issues/609